File: score_calculation/internal/goban/bouzy.go
package goban

const boardSize = 19

// InitializeInfluence initializes the board with influence values based on stone positions.
func InitializeInfluence(board [boardSize][boardSize]string) [boardSize][boardSize]int {
	influence := [boardSize][boardSize]int{}
	for y := 0; y < boardSize; y++ {
		for x := 0; x < boardSize; x++ {
			if board[y][x] == "B" {
				influence[y][x] = 64 // Positive high value for black
			} else if board[y][x] == "W" {
				influence[y][x] = -64 // Negative high value for white
			}
		}
	}
	return influence
}

// Dilation spreads influence across the board based on current positions and influence values.
func Dilation(board [boardSize][boardSize]string, influence *[boardSize][boardSize]int, steps int) {
	for step := 0; step < steps; step++ {
		tempInfluence := *influence
		for y := 0; y < boardSize; y++ {
			for x := 0; x < boardSize; x++ {
				spreadInfluence(x, y, board, &tempInfluence, (*influence)[y][x])
			}
		}
		*influence = tempInfluence
	}
}

// Erosion reduces influence based on the surrounding opposing or neutral influences.
func Erosion(influence *[boardSize][boardSize]int, steps int) {
	for step := 0; step < steps; step++ {
		tempInfluence := *influence
		for y := 0; y < boardSize; y++ {
			for x := 0; x < boardSize; x++ {
				if tempInfluence[y][x] != 0 {
					checkInfluenceRemoval(x, y, &tempInfluence, tempInfluence[y][x])
				}
			}
		}
		*influence = tempInfluence
	}
}

func spreadInfluence(x, y int, board [boardSize][boardSize]string, influence *[boardSize][boardSize]int, value int) {
	directions := [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	for _, d := range directions {
		nx, ny := x+d[0], y+d[1]
		if nx >= 0 && nx < boardSize && ny >= 0 && ny < boardSize && (board[ny][nx] == "0" || influence[ny][nx]*value >= 0) {
			influence[ny][nx] += value / 64 // Spread a fraction of the original value
		}
	}
}

func checkInfluenceRemoval(x, y int, influence *[boardSize][boardSize]int, value int) {
	directions := [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	countOpposite, countSame := 0, 0

	for _, d := range directions {
		nx, ny := x+d[0], y+d[1]
		if nx >= 0 && nx < boardSize && ny >= 0 && ny < boardSize {
			neighborValue := (*influence)[ny][nx]
			if neighborValue*value < 0 {
				countOpposite++
			} else if neighborValue*value > 0 {
				countSame++
			}
		}
	}

	if countOpposite > countSame { // Only erase if surrounded mostly by opposite influence
		(*influence)[y][x] = 0
	}
}

File: qnn/bouzy.py
import numpy as np

board_size = 19


def initialize_influence(board):
    influence = np.zeros((board_size, board_size), dtype=int)
    for y in range(board_size):
        for x in range(board_size):
            if board[y, x] == "B":
                influence[y, x] = 64  # Positive high value for black
            elif board[y, x] == "W":
                influence[y, x] = -64  # Negative high value for white
    return influence


def spread_influence(x, y, board, influence, value):
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    for d in directions:
        nx, ny = x + d[0], y + d[1]
        if 0 <= nx < board_size and 0 <= ny < board_size:
            if board[ny, nx] == "." or influence[ny, nx] * value >= 0:
                influence[ny, nx] += value // 64


def dilation(board, influence, steps):
    for _ in range(steps):
        temp_influence = influence.copy()
        for y in range(board_size):
            for x in range(board_size):
                spread_influence(x, y, board, temp_influence, influence[y, x])
        influence[:] = temp_influence


def check_influence_removal(x, y, influence, value):
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    count_opposite, count_same = 0, 0

    for d in directions:
        nx, ny = x + d[0], y + d[1]
        if 0 <= nx < board_size and 0 <= ny < board_size:
            neighbor_value = influence[ny, nx]
            if neighbor_value * value < 0:
                count_opposite += 1
            elif neighbor_value * value > 0:
                count_same += 1

    if count_opposite > count_same:
        influence[y, x] = 0


def erosion(influence, steps):
    for _ in range(steps):
        temp_influence = influence.copy()
        for y in range(board_size):
            for x in range(board_size):
                if temp_influence[y, x] != 0:
                    check_influence_removal(x, y, temp_influence, temp_influence[y, x])
        influence[:] = temp_influence
